src/main/java/me/jellysquid/mods/sodium/mixin/features/mipmaps/MixinMipmapHelper.java

```diff
@@ -0,0 +1,127 @@
+ * <p>This implementation fixes a number of issues with other implementations:</p>
+ *
+ * <li>
+ *     <ul>OptiFine blends in sRGB space, resulting in brightness losses.</ul>
+ *     <ul>Vanilla applies gamma correction to alpha values, which has weird results when alpha values aren't the same.</ul>
+ *     <ul>Vanilla computes a simple average of the 4 pixels, disregarding the relative alpha values of pixels. In
+ *         cutout textures, this results in a lot of pixels with high alpha values and dark colors, causing visual
+ *         artifacts.</ul>
+ * </li>
+ *
+ * This Mixin is ported from Iris at <a href="https://github.com/IrisShaders/Iris/blob/41095ac23ea0add664afd1b85c414d1f1ed94066/src/main/java/net/coderbot/iris/mixin/bettermipmaps/MixinMipmapGenerator.java">MixinMipmapGenerator</a>.
+ */
+@Mixin(MipmapHelper.class)
+public class MixinMipmapHelper {
+    // Generate some color tables for gamma correction.
+    private static final float[] SRGB_TO_LINEAR = new float[256];
+
+    static {
+        for (int i = 0; i < 256; i++) {
+            SRGB_TO_LINEAR[i] = (float) Math.pow(i / 255.0, 2.2);
+        }
+    }
+
+    /**
+     * @author coderbot
+     * @reason replace the vanilla blending function with our improved function
+     */
+    @Overwrite
+    private static int blend(int one, int two, int three, int four, boolean checkAlpha) {
+        // First blend horizontally, then blend vertically.
+        //
+        // This works well for the case where our change is the most impactful (grass side overlays)
+        return weightedAverageColor(weightedAverageColor(one, two), weightedAverageColor(three, four));
+    }
+
+    @Unique
+    private static int unpackAlpha(int color) {
+        return (color >>> 24) & 255;
+    }
+
+    @Unique
+    private static int weightedAverageColor(int one, int two) {
+        int alphaOne = unpackAlpha(one);
+        int alphaTwo = unpackAlpha(two);
+
+        // In the case where the alpha values of the same, we can get by with an unweighted average.
+        if (alphaOne == alphaTwo) {
+            return averageRgb(one, two, alphaOne);
+        }
+
+        // If one of our pixels is fully transparent, ignore it.
+        // We just take the value of the other pixel as-is. To compensate for not changing the color value, we
+        // divide the alpha value by 4 instead of 2.
+        if (alphaOne == 0) {
+            return (two & 0x00FFFFFF) | ((alphaTwo / 4) << 24);
+        }
+
+        if (alphaTwo == 0) {
+            return (one & 0x00FFFFFF) | ((alphaOne / 4) << 24);
+        }
+
+        // Use the alpha values to compute relative weights of each color.
+        float scale = 1.0f / (alphaOne + alphaTwo);
+        float relativeWeightOne = alphaOne * scale;
+        float relativeWeightTwo = alphaTwo * scale;
+
+        // Convert the color components into linear space, then multiply the corresponding weight.
+        float oneR = unpackLinearComponent(one, 0) * relativeWeightOne;
```

Comment:
We should use ColorABGR or ColorARGB here, they have documented unpack functions.